patterns.md

```diff
@@ -235,6 +235,20 @@ part of the domain then use it.
 
 Think of these modules as the _public_ API of the `$subcategory` of the domain.
 
+Sometimes it isn't obvious what the "category" of a function is. Use these
```

Comment:
Is this perhaps an oversimplification? I'm sceptical of treating source code purely as a filing cabinet (though it is partly that). Another important consideration is modularisation, which ultimately comes down to controlling dependencies.

An example: let's say we have a query that gets all accounts who have won the wheel of fortune. This would sit much better in a `wheel_of_fortune` package than in the `accounts` package. Ideally, you'd have a situation where `accounts` is entirely ignorant of the existence of `wheel_of_fortune`. Likewise a `spin_wheel` operation that operates on an account.

Perhaps I'm misunderstanding the advice, but it feels to me that it could lead us into a huge `accounts.queries` package.

The answer: I'm not sure, but I suspect at some point we'll need to formally decide on a Directed Acyclic Graph for the modules within `octoenergy.domain`, and that would be a better driver of where to locate code.

Reply:
Fair points.

It does say this immediately before the new section:

> This isn't a strict requirement - if there's a better naming convention for your
part of the domain then use it.

They are just rules-of-thumb for when there isn't a more suitable location. But we should improve the copy to make that clearer.

> An example: let's say we have a query that gets all accounts who have won the wheel of fortune. This would sit much better in a wheel_of_fortune package than in the accounts package. Ideally, you'd have a situation where accounts is entirely ignorant of the existence of wheel_of_fortune. Likewise a spin_wheel operation that operates on an account.

A good example and I agree that such functionality is too specific to one use-case to belong in the generic `domain.accounts.queries` package. I would argue that any functionality in a queries package should be agnostic of the use-cases that might use it. I would put such functionality in the use-case package itself but I suppose we might end up with feature-specific packages in the domain layer for this kind of thing, something like `domain.wheeloffortune.queries`. These would depend on the generic query/operations packages but not the other way around.

Following this structure should allow the directed graph of dependencies to emerge naturally. The graph would be:

```txt
use-case functionality 
    -> specific domain package 
        -> generic domain queries package
```

It's just a new layer in the lasagne really.